modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1702,264 +1719,549 @@ def tree_reduce(
+        new_index : list-like, optional
+            Index of the result. We may know this in advance, and if not provided it must be computed.
+        new_columns : list-like, optional
+            Columns of the result. We may know this in advance, and if not provided it must be computed.
+        apply_indices : list-like, optional
+            Labels of `axis ^ 1` to apply over (if `numeric_indices` is not specified).
+        numeric_indices : list-like, optional
+            Numeric indices to apply over (if `apply_indices` is not specified).
+        enumerate_partitions : bool, default: False
+            Whether pass partition index into applied `func` or not.
+            Note that `func` must be able to obtain `partition_idx` kwarg.
+        dtypes : pandas.Series, optional
+            Data types of the result. This is an optimization
+            because there are functions that always result in a particular data
+            type, and allows us to avoid (re)computing it.
+        copy_dtypes : bool, default: False
+            If True, the dtypes of the resulting dataframe are copied from the original,
+            and the ``dtypes`` argument is ignored.
 
         Returns
         -------
         PandasDataframe
-            A new PandasDataframe with the inferred schema.
+            New Modin DataFrame.
         """
-        # Compute dtypes on the specified columns, and then set those dtypes on a new frame
-        new_cols = self.take_2d_labels_or_positional(col_labels=col_labels)
-        new_cols_dtypes = new_cols.tree_reduce(0, pandas.DataFrame.infer_objects).dtypes
-        new_dtypes = self.dtypes.copy()
-        new_dtypes[col_labels] = new_cols_dtypes
-        return self.__constructor__(
+        if other is not None:
+            if not isinstance(other, list):
+                other = [other]
+            other = [o._partitions for o in other] if len(other) else None
+
+        if apply_indices is not None:
+            numeric_indices = self.axes[axis ^ 1].get_indexer_for(apply_indices)
+            apply_indices = self._get_dict_of_block_index(
+                axis ^ 1, numeric_indices
+            ).keys()
+
+        new_partitions = self._partition_mgr_cls.map_partitions_full_axis(
             self._partitions,
-            self._index_cache,
-            self._columns_cache,
-            self._row_lengths_cache,
-            self._column_widths_cache,
-            new_dtypes,
+            self._build_treereduce_func(axis, func),
+            axis=axis,
+            other_partitions=other,
+            keep_partitioning=True,
+            apply_indices=apply_indices,
+            enumerate_partitions=enumerate_partitions,
         )
+        # Index objects for new object creation. This is shorter than if..else
+        kw = self._make_init_labels_args(new_partitions, new_index, new_columns)
+        if copy_dtypes:
+            kw["dtypes"] = self._dtypes
+        elif isinstance(dtypes, type):
```

Comment:
judging by the method's signature we are only supposed to allow `pandas.Series` to be a `dtype` parameter, why is this logic here then? Let's either change the signature or adapt the logic somehow